File: tui_gateway/contracts/server_requests.py
"""Server→client requests: the backend asks the renderer a question (``server_requests.send``).

Every entry declares the params carried by its request frame (``session_id`` is added by the
transport) and the result returned by the client. ``request.cancel`` is the event-enveloped
withdrawal of an open request.
"""

from __future__ import annotations

from typing import Annotated, Literal, Union

from pydantic import Field

from .base import Params, Payload, Result, WireEnum
from .registry import event, server_request


class ServerRequestParams(Params):
    """Every request frame belongs to one gateway session."""

    session_id: str


class ValueResult(Result):
    """The answer to any one-string prompt (sudo, secret, vault prompts, desktop bridges):
    ``''`` means skipped / declined."""

    value: str


# ── clarify ───────────────────────────────────────────────────────────────────────────────────


class ClarifyQuestion(Params):
    qid: str
    question: str
    choices: list[str] | None
    multi_select: bool


class ClarifySingle(ServerRequestParams):
    """One clarify question; ``answer: ''`` is the client's explicit skip."""

    kind: Literal["single"]
    question: str
    choices: list[str] | None
    multi_select: bool


class ClarifyBatch(ServerRequestParams):
    """Multiple questions. ``answers`` is ``None`` when first sent and carries locked answers only
    on a reconnect replay."""

    kind: Literal["batch"]
    questions: list[ClarifyQuestion]
    answers: dict[str, str] | None


# ``server_request`` accepts this alias through the registry's TypeAdapter-backed declaration.
ClarifyRequestParams = Annotated[Union[ClarifySingle, ClarifyBatch], Field(discriminator="kind")]


class ClarifyAnswer(Result):
    """A single answer; ``answer: ''`` is an explicit skip."""

    answer: str


class ClarifyAnswers(Result):
    """The batch answer set. A deadline preserves locked answers with ``timed_out: true``."""

    answers: dict[str, str]
    timed_out: bool = False


ClarifyResult = Union[ClarifyAnswer, ClarifyAnswers]


# ``server_request`` accepts this union alias after A1's TypeAdapter registry change.
server_request("clarify", params=ClarifyRequestParams, result=ClarifyResult,  # type: ignore[arg-type]
               doc="The clarify tool: ask the user one question or a batch.")


# ── approval ──────────────────────────────────────────────────────────────────────────────────


class ApprovalChoice(WireEnum):
    once = "once"
    session = "session"
    always = "always"
    deny = "deny"


class ApprovalRequestParams(ServerRequestParams):
    """Closed payload from ``tools.approval`` after ``_approval_request_payload`` redacts the
    command and derives ``choices``."""

    request_id: str
    command: str
    description: str
    pattern_key: str | None
    pattern_keys: list[str] | None
    allow_permanent: bool | None
    allow_session: bool | None
    smart_denied: bool | None
    choices: list[ApprovalChoice]


class ApprovalResult(Result):
    choice: ApprovalChoice
    all: bool = False


server_request("approval", params=ApprovalRequestParams, result=ApprovalResult,
               doc="A dangerous command awaits the user's decision.")


# ── one-string prompts ────────────────────────────────────────────────────────────────────────


class EmptyRequestParams(ServerRequestParams):
    pass


class SudoRequestParams(ServerRequestParams):
    """Original command, redacted server-side before any password-injection rewrite."""

    command: str = ""


server_request("sudo", params=SudoRequestParams, result=ValueResult,
               doc="Masked sudo password for the terminal tool.")


class SecretMetadata(Params):
    """``tools.skills_tool_setup._capture_required_environment_variables`` supplies these keys."""

    skill_name: str
    help: str | None
    required_for: str | None


class SecretRequestParams(ServerRequestParams):
    env_var: str
    prompt: str
    metadata: SecretMetadata | None


server_request("secret", params=SecretRequestParams, result=ValueResult,
               doc="Masked value for a named env var in a skill setup flow.")


class VaultUnlockRequestParams(ServerRequestParams):
    backend: str
    display_name: str


server_request("vault.unlock_prompt", params=VaultUnlockRequestParams, result=ValueResult,
               doc="Master password to unlock an external password manager for this session.")


class VaultSaveLoginRequestParams(ServerRequestParams):
    origin: str
    site: str


server_request("vault.save_login", params=VaultSaveLoginRequestParams, result=ValueResult,
               doc="Save a login for a site; the value is JSON {identifier, password}.")


class VaultCodeRequestParams(ServerRequestParams):
    site: str | None
    hint: str | None


server_request("vault.code", params=VaultCodeRequestParams, result=ValueResult,
               doc="A one-time / 2FA code the user reads from their device.")


# ── desktop GUI bridges ───────────────────────────────────────────────────────────────────────


class ReadRangeRequestParams(ServerRequestParams):
    start: int | None
    count: int | None


server_request("terminal.read", params=ReadRangeRequestParams, result=ValueResult,
               doc="Read the visible in-app terminal buffer (JSON text answer).")
server_request("preview.read", params=ReadRangeRequestParams, result=ValueResult,
               doc="Read the in-app browser preview's text (JSON text answer).")
server_request("window.read", params=EmptyRequestParams, result=ValueResult,
               doc="Enumerate the native window below the app (JSON text answer).")


class PreviewActAction(WireEnum):
    elements = "elements"
    click = "click"
    hover = "hover"
    type = "type"
    scroll = "scroll"
    press = "press"
    strobe = "strobe"
    back = "back"
    forward = "forward"
    reload = "reload"
    pin = "pin"
    hold = "hold"
    unpin = "unpin"


class PreviewScrollTo(WireEnum):
    top = "top"
    bottom = "bottom"


class PreviewActRequestParams(ServerRequestParams):
    """Closed DOM-operation shape from ``tools.drive_preview_tool`` and
    ``tools.annotate_preview_tool``."""

    action: PreviewActAction
    ref: str | None
    selector: str | None
    text: str | None
    key: str | None
    submit: bool | None
    full: bool | None
    to: PreviewScrollTo | None
    amount: int | None
    max: int | None


server_request("preview.act", params=PreviewActRequestParams, result=ValueResult,
               doc="Click, type, scroll, or annotate inside the in-app browser preview.")


class TourAction(WireEnum):
    targets = "targets"
    show = "show"
    start = "start"
    next = "next"
    prev = "prev"
    stop = "stop"


class TourSurface(WireEnum):
    app = "app"
    preview = "preview"


class TourSide(WireEnum):
    top = "top"
    right = "right"
    bottom = "bottom"
    left = "left"


class TourStep(Params):
    """Closed DOM step from ``tools.tour_tool``'s ``_STEP_SCHEMA``."""

    selector: str | None
    title: str | None
    text: str | None
    side: TourSide | None


class TourRequestParams(ServerRequestParams):
    """Closed guided-tour DOM operation from ``tools.tour_tool``."""

    action: TourAction
    surface: TourSurface | None
    selector: str | None
    title: str | None
    text: str | None
    side: TourSide | None
    steps: list[TourStep] | None
    step_index: int | None


server_request("tour", params=TourRequestParams, result=ValueResult,
               doc="Drive a guided tour highlight in the desktop renderer.")


# ── withdrawal ────────────────────────────────────────────────────────────────────────────────


class WithdrawnRequestMethod(WireEnum):
    approval = "approval"
    clarify = "clarify"
    mcp_setup = "mcp.setup"
    preview_act = "preview.act"
    preview_read = "preview.read"
    secret = "secret"
    sudo = "sudo"
    terminal_read = "terminal.read"
    tour = "tour"
    vault_code = "vault.code"
    vault_save_login = "vault.save_login"
    vault_unlock_prompt = "vault.unlock_prompt"
    window_read = "window.read"


class RequestCancelReason(WireEnum):
    answered = "answered"
    interrupted = "interrupted"
    notify_failed = "notify_failed"
    shutdown = "shutdown"
    timeout = "timeout"


class RequestCancelPayload(Payload):
    id: str
    method: WithdrawnRequestMethod
    reason: RequestCancelReason


event("request.cancel", RequestCancelPayload,
      doc="The backend withdrew an open server→client request; clear the matching card only.")

File: apps/desktop/src/app/session/hooks/use-message-stream/gateway-event/server-requests.ts
import type { ClarifyParams, PreviewActParams, ServerRequestMap, TourParams, TourStep as WireTourStep } from '@hermes/shared'

import { type PreviewReadResult, readActivePreview } from '@/app/chat/right-rail/preview-reader'
import { readActiveTerminal, type TerminalReadResult } from '@/app/right-sidebar/terminal/buffer'
import { pendingClarifyToolPayload } from '@/app/session/hooks/use-session-actions/restore-pending-clarify'
import { translateNow } from '@/i18n'
import { restorePendingClarifyToolCall } from '@/lib/chat-messages'
import type { PreviewActAction, PreviewActResult } from '@/lib/preview-act/act-in-page'
import type { TourAction, TourResult, TourStep } from '@/lib/tour'
import { type ClarifyRequest, displayChoices, setClarifyRequest } from '@/store/clarify'
import type { ScopedServerRequest } from '@/store/gateway'
import { dispatchNativeNotification } from '@/store/native-notifications'
import {
  receiveApprovalRequest,
  setSecretRequest,
  setSudoRequest,
  setVaultCodeRequest,
  setVaultSaveLoginRequest,
  setVaultUnlockRequest
} from '@/store/prompts'
import { rememberServerRequest } from '@/store/server-requests'
import { requestScrollToBottom } from '@/store/thread-scroll'
import { $toursEnabled } from '@/store/tours'

import type { GatewayEventDeps } from './types'

/** The preview engine, loaded on demand so ~25KB of page-injectable source stays
 *  off the boot path (dev: a fresh copy per action so edits reach the guest — see
 *  the previous home of this loader in desktop-bridge.ts for the full story). */
const loadPreviewEngine = () => {
  const stable = () => import('@/app/chat/right-rail/preview-act')

  if (!import.meta.hot) {
    return stable().then(mod => mod.actOnActivePreview)
  }

  return import(/* @vite-ignore */ '/src/app/chat/right-rail/preview-act.ts?hot=' + Date.now())
    .catch(stable)
    // SAFETY: the hot URL serves the same module as `stable`, only cache-busted.
    .then(mod => mod.actOnActivePreview as Awaited<ReturnType<typeof stable>>['actOnActivePreview'])
}

export interface ServerRequestContext<M extends keyof ServerRequestMap = keyof ServerRequestMap> {
  deps: Pick<GatewayEventDeps, 'activeSessionIdRef' | 'sessionInterrupted' | 'updateSessionState' | 'upsertToolCall'>
  request: ScopedServerRequest<M>
  /** The session the request names. */
  sessionId: string
  /** The named session is the one on screen. */
  isActiveSession: boolean
}

type Handler<M extends keyof ServerRequestMap> = (ctx: ServerRequestContext<M>) => void

const markNeedsInput = (ctx: ServerRequestContext) => {
  if (ctx.sessionId) {
    ctx.deps.updateSessionState(ctx.sessionId, state => ({ ...state, needsInput: true }))
  }
}

const notifyInput = (ctx: ServerRequestContext, body: string) => {
  if (!ctx.request.replayed) {
    dispatchNativeNotification({
      body,
      kind: 'input',
      sessionId: ctx.sessionId || null,
      title: translateNow('notifications.native.inputTitle')
    })
  }
}

// ── Blocking-input family (clarify / approval / sudo / secret / vault / MCP setup) ──
// Every one is parked per-session (like clarify) so a BACKGROUND session's turn can
// raise it and wait — the sidebar flags "needs input" and the card surfaces once the
// user focuses that chat. The Python side blocks on the response frame.

/** The card to park for a clarify, or null when there is nothing to ask. */
const clarifyRequestFrom = (params: ClarifyParams, requestId: string, sessionId: string): ClarifyRequest | null => {
  const parked = { receivedAt: Date.now() / 1000, requestId, sessionId: sessionId || null }

  if (params.kind === 'batch') {
    // `answers` rides along only on a reconnect replay (locks the server already accepted).
    return params.questions.length > 0
      ? {
          ...parked,
          kind: 'batch',
          lockedAnswers: params.answers,
          questions: params.questions.map(question => ({ ...question, choices: displayChoices(question.choices) }))
        }
      : null
  }

  return params.question
    ? {
        ...parked,
        choices: displayChoices(params.choices),
        kind: 'single',
        multi_select: params.multi_select,
        question: params.question
      }
    : null
}

const clarify: Handler<'clarify'> = ctx => {
  const { deps, request, sessionId } = ctx

  if (sessionId && deps.sessionInterrupted(sessionId)) {
    request.respond({ answer: '' })

    return
  }

  const clarifyRequest = clarifyRequestFrom(request.params, request.id, sessionId)

  if (!clarifyRequest) {
    request.respond({ answer: '' })

    return
  }

  rememberServerRequest(request)
  setClarifyRequest(clarifyRequest)

  if (sessionId) {
    // A resumed/hydrated transcript may already contain this provider's clarify
    // call while carrying no live streamId. Re-arm that exact row instead of
    // letting the generic stream mutator append a second card.
    const occurredAt = Date.now() / 1000

    deps.updateSessionState(sessionId, state => {
      const projection = restorePendingClarifyToolCall(
        state.messages,
        pendingClarifyToolPayload(clarifyRequest),
        occurredAt
      )

      return {
        ...state,
        messages: projection.messages,
        streamId: projection.streamId,
        sawAssistantPayload: true,
        awaitingResponse: false,
        needsInput: true
      }
    })

    if (sessionId === deps.activeSessionIdRef.current) {
      requestScrollToBottom(sessionId)
    }
  }

  notifyInput(
    ctx,
    clarifyRequest.kind === 'batch'
      ? clarifyRequest.questions.map(question => question.question).join(' · ')
      : clarifyRequest.question
  )
}

const approval: Handler<'approval'> = ctx => {
  const { request, sessionId } = ctx
  const p = request.params
  const description = p.description || 'dangerous command'

  rememberServerRequest(request)
  void receiveApprovalRequest(null, {
    // false only when a tirith warning forbids it; backend omits the field otherwise.
    allowPermanent: p.allow_permanent !== false,
    choices: p.choices,
    command: p.command,
    description,
    // The approval queue's own id — `approval.pending` / `approval.received` / `approval.respond` key on it.
    requestId: p.request_id ?? undefined,
    serverRequestId: request.id,
    sessionId: sessionId || null,
    smartDenied: p.smart_denied === true
  }).catch(() => undefined)
  markNeedsInput(ctx)

  if (!request.replayed) {
    dispatchNativeNotification({
      actions: [
        {
          id: str(p.request_id) ? `approve:${str(p.request_id)}` : 'approve',
          text: translateNow('notifications.native.approveAction')
        },
        {
          id: str(p.request_id) ? `reject:${str(p.request_id)}` : 'reject',
          text: translateNow('notifications.native.rejectAction')
        }
      ],
      body: p.command || description,
      kind: 'approval',
      sessionId: sessionId || null,
      title: translateNow('notifications.native.approvalTitle')
    })
  }
}

const sudo: Handler<'sudo'> = ctx => {
  rememberServerRequest(ctx.request)
  setSudoRequest({
    command: ctx.request.params.command,
    requestId: ctx.request.id,
    sessionId: ctx.sessionId || null
  })
  markNeedsInput(ctx)
  notifyInput(ctx, translateNow('notifications.native.inputBody'))
}

const secret: Handler<'secret'> = ctx => {
  const { env_var: envVar, prompt } = ctx.request.params

  rememberServerRequest(ctx.request)
  setSecretRequest({ envVar, prompt, requestId: ctx.request.id, sessionId: ctx.sessionId || null })
  markNeedsInput(ctx)
  notifyInput(ctx, prompt || envVar || translateNow('notifications.native.inputBody'))
}

const vaultCode: Handler<'vault.code'> = ctx => {
  const site = ctx.request.params.site ?? ''

  rememberServerRequest(ctx.request)
  setVaultCodeRequest({
    hint: ctx.request.params.hint ?? '',
    requestId: ctx.request.id,
    sessionId: ctx.sessionId || null,
    site
  })
  markNeedsInput(ctx)
  notifyInput(ctx, translateNow('prompts.vaultCodeTitle', site))
}

const vaultSaveLogin: Handler<'vault.save_login'> = ctx => {
  const { origin } = ctx.request.params
  const site = ctx.request.params.site || origin

  rememberServerRequest(ctx.request)
  setVaultSaveLoginRequest({ origin, requestId: ctx.request.id, sessionId: ctx.sessionId || null, site })
  markNeedsInput(ctx)
  notifyInput(ctx, translateNow('prompts.vaultSaveTitle', site))
}

const vaultUnlockPrompt: Handler<'vault.unlock_prompt'> = ctx => {
  const { backend } = ctx.request.params
  const displayName = ctx.request.params.display_name || backend

  rememberServerRequest(ctx.request)
  setVaultUnlockRequest({ backend, displayName, requestId: ctx.request.id, sessionId: ctx.sessionId || null })
  markNeedsInput(ctx)
  notifyInput(ctx, translateNow('prompts.vaultUnlockTitle', displayName))
}

// ── Desktop-surface bridges (answered immediately, no card) ─────────────────

type SurfaceRequest = ScopedServerRequest<'preview.act' | 'preview.read' | 'terminal.read' | 'tour' | 'window.read'>

interface RefusedAction {
  error: string
  success: false
}

type WindowBelow = Awaited<ReturnType<NonNullable<NonNullable<typeof window.hermesDesktop>['readWindowBelow']>>>

type SurfaceAnswer = null | PreviewActResult | PreviewReadResult | RefusedAction | TerminalReadResult | TourResult | WindowBelow

/** Answer a string-valued request with a JSON-encoded result ('' = nothing / unavailable). */
const answerValue = (request: SurfaceRequest, result: SurfaceAnswer) =>
  request.respond({ value: result ? JSON.stringify(result) : '' })

const refused = (error: string): RefusedAction => ({ error, success: false })

const failureText = (error: Error | string) => (error instanceof Error ? error.message : String(error))

const terminalRead: Handler<'terminal.read'> = ({ request }) => {
  // read_terminal tool: serialize the renderer's xterm buffer. Empty = no live pane.
  const { count, start } = request.params

  answerValue(request, readActiveTerminal({ count: count ?? undefined, start: start ?? undefined }))
}

const previewRead: Handler<'preview.read'> = ({ request }) => {
  // read_preview tool: the active preview tab's page text is async. Empty = nothing open.
  const { count, start } = request.params

  void readActivePreview({ count: count ?? undefined, start: start ?? undefined }).then(result =>
    answerValue(request, result)
  )
}

/** The wire operation as the preview engine takes it (nav verbs included; the engine routes those itself). */
const previewAction = (p: PreviewActParams): Omit<PreviewActAction, 'kind'> & { kind: PreviewActParams['action'] } => ({
  amount: p.amount ?? undefined,
  full: p.full ?? undefined,
  key: p.key ?? undefined,
  kind: p.action,
  max: p.max ?? undefined,
  ref: p.ref ?? undefined,
  selector: p.selector ?? undefined,
  submit: p.submit ?? undefined,
  text: p.text ?? undefined,
  to: p.to ?? undefined
})

const previewAct: Handler<'preview.act'> = ({ isActiveSession, request, sessionId }) => {
  // drive_preview tool: click/type/scroll/press inside the guest page. Active
  // session only: a background turn must never reach into the page the user is
  // working in (desktop AGENTS.md: offer, don't hijack). Every mounted window can
  // observe the same request; a scoped mismatch belongs to another window, so
  // answering here would race the owner — stay silent.
  if (sessionId && !isActiveSession) {
    return
  }

  if (!isActiveSession) {
    answerValue(request, refused('The in-app browser only takes actions in the session the user is looking at.'))

    return
  }

  void loadPreviewEngine()
    .then(run => run(previewAction(request.params)))
    .then(
      result => answerValue(request, result),
      error => answerValue(request, refused(failureText(error)))
    )
}

const windowRead: Handler<'window.read'> = ({ request }) => {
  // read_window_below tool: main owns native window enumeration. Empty =
  // unavailable (older shell without the handler, Wayland, …) — without an
  // answer the tool would stall its full 30s deadline.
  const read = window.hermesDesktop?.readWindowBelow

  void Promise.resolve(read ? read() : null).then(
    result => answerValue(request, result),
    () => answerValue(request, null)
  )
}

const tourStep = (step: WireTourStep): TourStep => ({
  selector: step.selector ?? undefined,
  side: step.side ?? undefined,
  text: step.text ?? undefined,
  title: step.title ?? undefined
})

const tourAction = (p: TourParams): TourAction => ({
  ...tourStep(p),
  kind: p.action,
  startAt: p.step_index ?? undefined,
  steps: p.steps?.map(tourStep)
})

const tour: Handler<'tour'> = ({ isActiveSession, request, sessionId }) => {
  // tour tool: one guided-tour action via driver.js, app DOM or preview guest
  // page. Active session only, same window-ownership rule as preview.act.
  if (sessionId && !isActiveSession) {
    return
  }

  if (!$toursEnabled.get()) {
    // Refused in words, not silently dropped: a no-op would leave the agent
    // narrating a spotlight the user can't see.
    answerValue(request, refused('The user has turned guided tours off.'))

    return
  }

  if (!isActiveSession) {
    answerValue(request, refused('Tours only run in the session the user is looking at.'))

    return
  }

  void import('@/lib/tour')
    .then(({ runTour }) => runTour(tourAction(request.params), request.params.surface ?? 'app'))
    .then(
      result => answerValue(request, result),
      error => answerValue(request, refused(failureText(error)))
    )
}

/** Method → handler. tsc holds this to every `ServerRequestMap` key. */
const HANDLERS: { [M in keyof ServerRequestMap]: Handler<M> } = {
  approval,
  clarify,
  'preview.act': previewAct,
  'preview.read': previewRead,
  secret,
  sudo,
  'terminal.read': terminalRead,
  tour,
  'vault.code': vaultCode,
  'vault.save_login': vaultSaveLogin,
  'vault.unlock_prompt': vaultUnlockPrompt,
  'window.read': windowRead
}

/** Dispatch one server→client request to its typed handler. */
export function handleServerRequest<M extends keyof ServerRequestMap>(
  request: ScopedServerRequest<M>,
  deps: ServerRequestContext['deps'],
  activeSessionId: null | string
): void {
  const sessionId = request.params.session_id

  HANDLERS[request.method]({
    deps,
    request,
    sessionId,
    isActiveSession: Boolean(sessionId) && sessionId === activeSessionId
  })
}
